File: webapp/packages/grounding/src/space.ts
/**
 * Core abstractions for symbol grounding framework.
 *
 * Extracted from Coq formal specifications in coq/Color/Core.v
 */

import type { OKLab } from '@oklab/core';

// Type for symbols (string identifiers)
export type Symbol = string;

/**
 * Abstract ground space for symbol grounding.
 *
 * Corresponds to Coq Record GroundSpace with carrier, distance, mix, validate.
 */
export interface Space<S> {
  /** Distance metric between points in the space */
  distance(a: S, b: S): number;

  /** Convex combination of points with weights */
  mix(points: S[], weights: number[]): S;

  /** Check if point is valid in this space */
  validate(point: S): boolean;
}

/**
 * Abstract geometric region in a ground space.
 *
 * Corresponds to Coq Record GroundRegion with contains predicate.
 */
export interface GroundRegion<S> {
  /** Check if point is contained in this region */
  contains(point: S): boolean;
}

/**
 * Mapping from symbols to regions in a ground space.
 *
 * Corresponds to Coq Record Grounding with regions function.
 */
export class Grounding<S> {
  private regions = new Map<Symbol, GroundRegion<S>>();

  constructor(private space: Space<S>) {}

  /** Bind a symbol to a region in the space */
  bindRegion(symbol: Symbol, region: GroundRegion<S>): void {
    this.regions.set(symbol, region);
  }

  /** Get the region for a symbol */
  getRegion(symbol: Symbol): GroundRegion<S> {
    const region = this.regions.get(symbol);
    if (!region) {
      throw new Error(`No region bound for symbol: ${symbol}`);
    }
    return region;
  }

  /** Find the symbol whose region is closest to the point */
  nearestSymbol(point: S): Symbol | null {
    let minDistance = Infinity;
    let nearest: Symbol | null = null;

    for (const [symbol, region] of this.regions) {
      // Check if point is in region
      if (region.contains(point)) {
        return symbol;
      }

      // Fallback: find closest region center
      // For SphericalRegion, use the center property
      const center = this._getRegionCenter(region);
      if (center) {
        const distance = this.space.distance(point, center);
        if (distance < minDistance) {
          minDistance = distance;
          nearest = symbol;
        }
      }
    }

    return nearest;
  }

  /** Calculate similarity between two symbols' regions */
  similarity(symbol1: Symbol, symbol2: Symbol): number {
    const region1 = this.getRegion(symbol1);
    const region2 = this.getRegion(symbol2);

    const center1 = this._getRegionCenter(region1);
    const center2 = this._getRegionCenter(region2);

    if (center1 && center2) {
      // Similarity based on distance (higher similarity = lower distance)
      const distance = this.space.distance(center1, center2);
      return 1.0 / (1.0 + distance);
    }

    // No center information available
    return 0.0;
  }

  /** Helper to extract center from region (type-safe) */
  private _getRegionCenter(region: GroundRegion<S>): S | null {
    // Check if region has a center property (like SphericalRegion)
    if ('center' in region && typeof (region as any).center !== 'undefined') {
      return (region as any).center as S;
    }
    return null;
  }
}

File: oklab_grounding/space.py
"""
Core abstractions for symbol grounding framework.

Extracted from Coq formal specifications in coq/Color/Core.v
"""

from abc import ABC, abstractmethod
from typing import TypeVar, Generic, Dict, List, Protocol
import math

# Type variable for points in the space
S = TypeVar('S')

# Symbol as string identifier (matching Coq Symbol := string)
Symbol = str

class GroundSpace(ABC, Generic[S]):
    """
    Abstract ground space for symbol grounding.

    Corresponds to Coq Record GroundSpace with carrier, distance, mix, validate.
    """

    @abstractmethod
    def distance(self, a: S, b: S) -> float:
        """Distance metric between points in the space."""
        pass

    @abstractmethod
    def mix(self, points: List[S], weights: List[float]) -> S:
        """Convex combination of points with weights."""
        pass

    @abstractmethod
    def validate(self, point: S) -> bool:
        """Check if point is valid in this space."""
        pass

class GroundRegion(ABC, Generic[S]):
    """
    Abstract geometric region in a ground space.

    Corresponds to Coq Record GroundRegion with contains predicate.
    """

    @abstractmethod
    def contains(self, point: S) -> bool:
        """Check if point is contained in this region."""
        pass

class Grounding(Generic[S]):
    """
    Mapping from symbols to regions in a ground space.

    Corresponds to Coq Record Grounding with regions function.
    """

    def __init__(self, space: GroundSpace[S]):
        self.space = space
        self._regions: Dict[Symbol, GroundRegion[S]] = {}

    def bind_region(self, symbol: Symbol, region: GroundRegion[S]) -> None:
        """Bind a symbol to a region in the space."""
        self._regions[symbol] = region

    def get_region(self, symbol: Symbol) -> GroundRegion[S]:
        """Get the region for a symbol."""
        return self._regions[symbol]

    def nearest_symbol(self, point: S) -> Symbol:
        """Find the symbol whose region is closest to the point."""
        if not self._regions:
            raise ValueError("No regions bound")

        min_distance = float('inf')
        nearest = None

        for symbol, region in self._regions.items():
            # Simple center-based distance for now
            # In practice, would need region-specific distance calculation
            if region.contains(point):
                return symbol
            # Placeholder: assume regions have a center method
            # This needs to be extended based on concrete region types
            try:
                center = getattr(region, 'center', lambda: point)()
                dist = self.space.distance(point, center)
                if dist < min_distance:
                    min_distance = dist
                    nearest = symbol
            except AttributeError:
                continue

        return nearest or list(self._regions.keys())[0]

    def similarity(self, symbol1: Symbol, symbol2: Symbol) -> float:
        """Calculate similarity between two symbols' regions."""
        region1 = self.get_region(symbol1)
        region2 = self.get_region(symbol2)

        # Placeholder implementation
        # In practice, would compute region overlap or distance
        try:
            center1 = getattr(region1, 'center', lambda: None)()
            center2 = getattr(region2, 'center', lambda: None)()
            if center1 is not None and center2 is not None:
                return 1.0 / (1.0 + self.space.distance(center1, center2))
        except AttributeError:
            pass

        return 0.0  # No similarity information available